google/cloud/firestore_bundle/bundle.py

```diff
@@ -0,0 +1,439 @@
+                read_time=self._doc_metadata_map[
+                    bundle_element.document.name
+                ].read_time,
+                create_time=bundle_element.document.create_time,  # type: ignore
+                update_time=bundle_element.document.update_time,  # type: ignore
+            )
+            self.add_document(snapshot)
+
+            bundled_document = self.documents.get(snapshot.reference._document_path)
+            for query_name in self._doc_metadata_map[
+                bundle_element.document.name
+            ].queries:
+                bundled_document.metadata.queries.append(query_name)  # type: ignore
+        else:
+            raise ValueError(f"Unexpected type of BundleElement: {type}")
+
+    def build(self) -> str:
+        """Iterates over the bundle's stored documents and queries and produces
+        a single length-prefixed json string suitable for long-term storage.
+
+        Example:
+
+            from google.cloud import firestore
+
+            db = firestore.Client()
+            collection_ref = db.collection(u'users')
+
+            bundle = firestore.FirestoreBundle('my bundle')
+            bundle.add('app-users', collection_ref._query())
+
+            serialized_bundle: str = bundle.build()
+
+            # Now upload `serialized_bundle` to Google Cloud Storage, store it
+            # in Memorystore, or any other storage solution.
+
+        Returns:
+            str: The length-prefixed string representation of this bundle'
+                contents.
+        """
+        buffer: str = ""
+
+        named_query: NamedQuery
+        for named_query in self.named_queries.values():
+            buffer += self._compile_bundle_element(
+                BundleElement(named_query=named_query)
+            )
+
+        bundled_document: "_BundledDocument"  # type: ignore
+        document_count: int = 0
+        for bundled_document in self.documents.values():
+            buffer += self._compile_bundle_element(
+                BundleElement(document_metadata=bundled_document.metadata)
+            )
+            document_count += 1
+            buffer += self._compile_bundle_element(
+                BundleElement(document=bundled_document.snapshot._to_protobuf()._pb,)
+            )
+
+        metadata: BundleElement = BundleElement(
+            metadata=self._deserialized_metadata
+            or BundleMetadata(
+                id=self.name,
+                create_time=_helpers.build_timestamp(),
+                version=FirestoreBundle.BUNDLE_SCHEMA_VERSION,
+                total_documents=document_count,
+                total_bytes=len(buffer.encode("utf-8")),
+            )
+        )
+        return f"{self._compile_bundle_element(metadata)}{buffer}"
+
+    def _compile_bundle_element(self, bundle_element: BundleElement) -> str:
+        serialized_be: str = json.dumps(BundleElement.to_dict(bundle_element))
+        return f"{len(serialized_be)}{serialized_be}"
```

Comment:
Does `len` return the length in bytes? Or is it the length of the JSON in String form? It should be bytes.

Reply:
my quick stack overflow search suggests `len(s.encode('utf-8'))`..

Reply:
I had that same suspicion and used that first, and even included compound emojis, but it never seemed to return a different answer. But, I'll revisit this.

Reply:
I just double checked this and confirmed that `json.dumps` encodes non-ASCII characters.